### PR TITLE
MAISTRA-1921: Support dynamic crypto based on environment

### PR DESCRIPTION
**Pull request by knrc**
This PR modifies the defaults used when running within a FIPS environment.

### CODE REVIEW

**Review by jwendell — APPROVED** [review]

**Review by brian-avery — APPROVED** [review]

File: source/extensions/transport_sockets/tls/context_config_impl.cc [hunk 1]
```diff
@@ -19,6 +21,30 @@ namespace Tls {
 
 namespace {
 
+absl::once_flag fipsOnce;
+bool fipsEnabled;
+
+void initFipsEnabled() {
+  std::ifstream file("/proc/sys/crypto/fips_enabled");
+  if (file.fail()) {
+    std::cout << "FIPS " << (fipsEnabled ? "enabled" : "disabled") << "\n";
```
Comment (jwendell) [inline source/extensions/transport_sockets/tls/context_config_impl.cc:30]: FIPS is always disabled here. What's the content of `fipsEnabled` variable at this point? Better to go back to a static message?

File: source/extensions/transport_sockets/tls/context_config_impl.cc [hunk 2]
```diff
@@ -19,6 +21,30 @@ namespace Tls {
...
+
+    std::string fipsEnabledText = file_string.str();
+    fipsEnabledText.erase(fipsEnabledText.find_last_not_of("\n") + 1);
+    fipsEnabled = fipsEnabledText.compare("1") == 0;
+    std::cout << "FIPS " << (fipsEnabled ? "enabled" : "disabled") << ": " << fipsEnabledText << "\n";
+  }
+}
+
+bool isFipsEnabled() {
+  absl::call_once(fipsOnce, initFipsEnabled);
```
Comment (jwendell) [inline source/extensions/transport_sockets/tls/context_config_impl.cc:44]: Tricky :) In the old good days we would just use a `static` variable that would store whether or not we need to run something :)

File: source/extensions/transport_sockets/tls/context_config_impl.cc (hunk 1 above)
Comment (knrc) [reply source/extensions/transport_sockets/tls/context_config_impl.cc:30]: True, this was a static message and should be reverted to that

File: source/extensions/transport_sockets/tls/context_config_impl.cc (hunk 2 above)
Comment (knrc) [reply source/extensions/transport_sockets/tls/context_config_impl.cc:44]: It seemed safer than playing directly with mutexes :)

Comment (knrc) [reply source/extensions/transport_sockets/tls/context_config_impl.cc:44]: and statics don't work in certain memory models and regularly break

File: source/extensions/transport_sockets/tls/context_config_impl.cc [hunk 3]
```diff
@@ -282,7 +308,24 @@ unsigned ContextConfigImpl::tlsVersionFromProto(
 const unsigned ClientContextConfigImpl::DEFAULT_MIN_VERSION = TLS1_VERSION;
 const unsigned ClientContextConfigImpl::DEFAULT_MAX_VERSION = TLS1_2_VERSION;
 
-const std::string ClientContextConfigImpl::DEFAULT_CIPHER_SUITES =
+// FIPS configuration
+const std::string ClientContextConfigImpl::DEFAULT_FIPS_CIPHER_SUITES =
```
Comment (dmitri-d) [inline source/extensions/transport_sockets/tls/context_config_impl.cc:312]: Both server and client default cipher suites look the same. I think they are going to stay this way too. Any reason not to move these to the base class then?

File: source/extensions/transport_sockets/tls/context_config_impl.cc [hunk 4]
```diff
@@ -19,6 +21,30 @@ namespace Tls {
 
 namespace {
 
+absl::once_flag fipsOnce;
+bool fipsEnabled;
+
+void initFipsEnabled() {
+  std::ifstream file("/proc/sys/crypto/fips_enabled");
+  if (file.fail()) {
+    std::cout << "FIPS disabled\n";
```
Comment (dmitri-d) [inline source/extensions/transport_sockets/tls/context_config_impl.cc:30]: Could you replace cout's with `ENVOY_LOG(debug, "message here")`? You'll need to include `"common/common/logger.h"` and add `"//source/common/common:logger_lib"` dependency to the `context_lib` target in `source/extensions/transport_sockets/tls/BUILD`.

Comment (knrc) [reply source/extensions/transport_sockets/tls/context_config_impl.cc:30]: No, I had that initially and it's at warning level so we don't see it.

Comment (knrc) [reply source/extensions/transport_sockets/tls/context_config_impl.cc:30]: I don't want to set it at warning, or higher, to guarantee it is displayed every time

File: source/extensions/transport_sockets/tls/context_config_impl.cc (hunk 3 above)
Comment (knrc) [reply source/extensions/transport_sockets/tls/context_config_impl.cc:312]: With the exception of the TLS max they are but this just follows the upstream behaviour which I'm not changing

**Comment by knrc** [comment]
I'm going to revert the code tomorrow, back to the logging version, and then try to bypass the level settings to always display

File: source/extensions/transport_sockets/tls/context_config_impl.cc (hunk 4 above)
Comment (dmitri-d) [reply source/extensions/transport_sockets/tls/context_config_impl.cc:30]: Would info level work here?

**Comment by knrc** [comment]
I had a long think about this while on a bike ride

The following options have been tried
- variable initialiser with logging framework (impacts some fragile Envoy tests which break with the extra output since logging levels have not been set to warning)
- on demand initialisation with logging framework (logging level is at warning so no output)
- on demand initialisation with cout - Success but not ideal solution

since each of these causes issues, or is undesirable in some way, I've decided to revert to the original variable initialisation and not log the status.

**Comment by knrc** [comment]
Test failure is IpVersions/IntegrationTest.ResponseFramedByConnectionCloseWithReadLimits/IPv4 which seems to be a flaky test, it has happened with previous PRs.

**Comment by jwendell** [comment]
> I had a long think about this while on a bike ride
> 
> The following options have been tried
> 
>     * variable initialiser with logging framework (impacts some fragile Envoy tests which break with the extra output since logging levels have not been set to warning)
> 
>     * on demand initialisation with logging framework (logging level is at warning so no output)
> 
>     * on demand initialisation with cout - Success but not ideal solution
> 
> 
> since each of these causes issues, or is undesirable in some way, I've decided to revert to the original variable initialisation and not log the status.

If you wish, you can append something like "/FIPS" in the output of `--version` command here: https://github.com/maistra/envoy/blob/maistra-1.1/source/common/common/version.cc#L22

**Comment by knrc** [comment]
@jwendell that's a good idea, although at this point I've spent more time on fixing the logging than I did on fixing the actual issue

File: source/extensions/transport_sockets/tls/context_config_impl.cc (hunk 4 above)
Comment (knrc) [reply source/extensions/transport_sockets/tls/context_config_impl.cc]: No, that's what I had but the default settings are for warning and above.  @jwendell came up with a better solution and I'll try to implement that in a separate PR.

**Pull request merged**